templates/cicd/pipeline_cfn.yml

```diff
@@ -234,16 +250,27 @@ Resources:
               - Name: SCCheckoutArtifact
             OutputArtifacts:
               - Name: BuildOutput
-{{$length := len .Stages}}{{if gt $length 0}}{{range $stage := .Stages}}{{$svcNum := len $stage.LocalServices}}{{if gt $svcNum 0}}
+{{$length := len .Stages}}{{if gt $length 0}}{{range $index, $stage := .Stages}}{{$svcNum := len $stage.LocalServices}}{{if gt $svcNum 0}}
         - Name: DeployTo-{{$stage.Name}}
           Actions:{{if $stage.Prod}}
             - Name: ApprovePromotionTo-{{$stage.Name}}
               ActionTypeId:
                 Category: Approval
                 Owner: AWS
-                Provider: Manual
                 Version: 1
+                Provider: Manual
               RunOrder: 1{{end}}{{range $svc := $stage.LocalServices}}
+            - Name: TestCommands{{if $stage.TestCommands}}
```

Comment:
Is it okay for 
```yaml
            - Name: TestCommands
              InputArtifacts:
                - Name: SCCheckoutArtifact
```
this to be empty if there are no test commands?

Reply:
Thank you. Fixed. Here's an example of the generated template when one deploy stage has test commands and one does not: 
```
BuildTestCommands0:
    Type: AWS::CodeBuild::Project
    Properties:
      Name: BuildTestCommands0
      EncryptionKey: !ImportValue sample-ArtifactKey
      ServiceRole: !GetAtt BuildProjectRole.Arn
      Artifacts:
        Type: NO_ARTIFACTS
      Environment:
        Type: LINUX_CONTAINER
        Image: aws/codebuild/amazonlinux2-x86_64-standard:3.0
        ComputeType: BUILD_GENERAL1_SMALL
      Source:
        Type: NO_SOURCE
        BuildSpec: "version: 0.2\nphases:\n  build:\n    commands: []"
  BuildTestCommands1:
    Type: AWS::CodeBuild::Project
    Properties:
      Name: BuildTestCommands1
      EncryptionKey: !ImportValue sample-ArtifactKey
      ServiceRole: !GetAtt BuildProjectRole.Arn
      Artifacts:
        Type: NO_ARTIFACTS
      Environment:
        Type: LINUX_CONTAINER
        Image: aws/codebuild/amazonlinux2-x86_64-standard:3.0
        ComputeType: BUILD_GENERAL1_SMALL
      Source:
        Type: NO_SOURCE
        BuildSpec: "version: 0.2\nphases:\n  build:\n    commands: [\"echo 'next test'\"]"
```
and only the second BuildTestCommands is called:
```
- Name: DeployTo-test
          Actions:
          
            - Name: CreateOrUpdate-svc1-test
              Region: us-west-2
              ActionTypeId:
                Category: Deploy
                Owner: AWS
                Version: 1
                Provider: CloudFormation
              Configuration:
                # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-action-reference.html
                ChangeSetName: sample-test-svc1
                ActionMode: CREATE_UPDATE
                StackName: sample-test-svc1
                Capabilities: CAPABILITY_NAMED_IAM
                TemplatePath: BuildOutput::infrastructure/svc1.stack.yml
                TemplateConfiguration: BuildOutput::infrastructure/svc1-test.params.json
                # The ARN of the IAM role (in the env account) that
                # AWS CloudFormation assumes when it operates on resources
                # in a stack in an environment account.
                RoleArn: arn:aws:iam::327148513104:role/sample-test-CFNExecutionRole
              InputArtifacts:
                - Name: BuildOutput
              RunOrder: 2
              # The ARN of the environment manager IAM role (in the env
              # account) that performs the declared action. This is assumed
              # through the roleArn for the pipeline.
              RoleArn: arn:aws:iam::327148513104:role/sample-test-EnvManagerRole
        - Name: DeployTo-prod
          Actions:
          
            - Name: TestCommands
              ActionTypeId:
                Category: Test
                Owner: AWS
                Version: 1
                Provider: CodeBuild
              Configuration:
                ProjectName: !Ref BuildTestCommands1
              RunOrder: 1
              InputArtifacts:
                - Name: SCCheckoutArtifact
```